[HTML5 Article Renderer] Apply table styling

WinnyChang opened the pull request:
## Summary
- Applied relevant styles to all table elements.
- Enabled horizontal scrolling for tables on smaller screens.
- Added a visible focus outline to the table container when focused.

Note: Temporary image styles have been applied to prevent horizontal page stretching.

### Screenshots - Desktop

#### Before
![before-table-styling-desktop](https://github.com/user-attachments/assets/2119acec-0dac-43a1-a830-bf1f25613483)

#### After
![after-table-styling-desktop](https://github.com/user-attachments/assets/040196e8-bd6f-490b-ab98-460d6d374788)

### Screenshots - Mobile (Pixel 7)

#### Before
<img width="30%" alt="before-table-styling-mobile" src="https://github.com/user-attachments/assets/997b5dfa-bcbe-4887-8b41-e0b1936b8b03" />

#### After
<img width="30%" alt="after-table-styling-mobile" src="https://github.com/user-attachments/assets/d60557b8-17f6-406a-a9bc-022c87c9d9a4" />

### Screenshot - Focus outline
<img width="50%" alt="after-table-styling-focus" src="https://github.com/user-attachments/assets/bf26feae-94d5-4109-af3d-37e901eb5b4e" />

### Screen Recordings
https://github.com/user-attachments/assets/47c84987-a82d-4ae5-ab03-9e8c2c17b3a4

https://github.com/user-attachments/assets/ded73333-243d-4c05-a712-9e2fefbf290d

https://github.com/user-attachments/assets/f32bd70c-eb46-4268-8611-26809ee5544e

## References
Issue - [#13482](https://github.com/learningequality/kolibri/issues/13482)

## Reviewer guidance
Check if:
- All styled table elements visually match the Figma specification.
- Tables become horizontally scrollable on small screens when their width exceeds the viewport, without breaking the layout or causing horizontal page scrolling.
- Tables are keyboard-tabbable when scrollable, with a visible focus outline.

WinnyChang committed:
Apply basic table styles

WinnyChang committed:
Wrap tables in responsive container for horizontal scrolling

WinnyChang committed:
Dynamically set table width based on column count

WinnyChang committed:
Add focus outline to table container

AllanOXDi commented on packages/kolibri-common/components/SafeHTML/index.js:
```diff
@@ -31,12 +31,28 @@ export default {
+          const colCount = firstRow ? firstRow.children.length : 0;
+          let tableWidth = '640px';
+          if (colCount > 3) {
+            tableWidth = `${colCount * 200}px`;
```
I am assuming that the  hardcoded width calculation (colCount * 200px) might not work well for all table content types

WinnyChang replied:
Do you mean the tables that have a different number of columns on different rows?

AllanOXDi commented on packages/kolibri-common/components/SafeHTML/index.js:
```diff
@@ -31,12 +31,28 @@ export default {
         attributes.class = 'safe-html';
-        return h(node.tagName.toLowerCase(), { attrs: attributes }, mapChildren(node.childNodes));
+        if (tag === 'table') {
+          const firstRow = node.querySelector('tr');
```
Seems like you  are trying to access firstRow.children.length, which would throw an error if firstRow is null. if there's no <tr> at all in the table: firstRow will be null. Maybe we could try
```
const firstRow = node.querySelector('tr');
const colCount = firstRow ? firstRow.children.length : 0;
```

WinnyChang replied:
Thanks for pointing that out! The code you shared looks the same as what I’m currently using, so I just wanted to double-check—was there a typo or something I might have overlooked?

akolson commented:
@WinnyChang I think it makes sense to have the footer rows complete with no spanning applied as seen below;
![image](https://github.com/user-attachments/assets/9dc44095-e700-41ac-92de-1fba62e9c269)

akolson commented:
> I couldn't figure out how to use the [$coreOutline](https://design-system.learningequality.org/accessibility/#keyboard), so I hard-coded the focus outline styles for now. Is there a better way to do this?

I think there should be a way to do to. Let's see if we can find you some code examples on how it is used.

WinnyChang commented:
> @WinnyChang I think it makes sense to have the footer rows complete with no spanning applied as seen below; ![image](https://private-user-images.githubusercontent.com/5203639/461237885-9dc44095-e700-41ac-92de-1fba62e9c269.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0MDUzMDYsIm5iZiI6MTc1MTQwNTAwNiwicGF0aCI6Ii81MjAzNjM5LzQ2MTIzNzg4NS05ZGM0NDA5NS1lNzAwLTQxYWMtOTJkZS0xZmJhNjJlOWMyNjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDFUMjEyMzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzZkNGM5NmZlODYyY2JmMzQwYTI5ZjkzYTBjMmY1NmNhOGJhZGRhZjJmMjYyY2RkODFlZmY4NTNmZmEyODE2ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hcUZ-SwsYzSGdnCzBm1C_Myl4_Eb3vbU4nz2GLQRVB0)

The table in the test resource has the following structure:
```
<table data-v-68300410="" class="safe-html" style="width: 640px;">
  <caption data-v-68300410="" class="safe-html">Annual Migration Distances by Species</caption>
  <thead data-v-68300410="" class="safe-html">
    <tr data-v-68300410="" class="safe-html">
      <th data-v-68300410="" class="safe-html">Species</th>
      <th data-v-68300410="" class="safe-html">Distance (miles)</th>
      <th data-v-68300410="" class="safe-html">Route</th>
    </tr>
  </thead>
  <tbody data-v-68300410="" class="safe-html">
    <tr data-v-68300410="" class="safe-html">
      <td data-v-68300410="" class="safe-html">Ruby-throated</td>
      <td data-v-68300410="" class="safe-html">500-2,000</td>
      <td data-v-68300410="" class="safe-html">Eastern US to Central America</td>
    </tr>
    <tr data-v-68300410="" class="safe-html">
      <td data-v-68300410="" class="safe-html">Rufous</td>
      <td data-v-68300410="" class="safe-html">3,900</td>
      <td data-v-68300410="" class="safe-html">Alaska to Mexico</td>
    </tr>
  </tbody>
  <tfoot data-v-68300410="" class="safe-html">
    <tr data-v-68300410="" class="safe-html">
      <td data-v-68300410="" class="safe-html">Some hummingbirds cross the Gulf of Mexico in a single 18-20 hour flight!</td>
    </tr>
  </tfoot>
</table>
```
Since there's only one `td` in the `tfoot`, I couldn't style those empty spaces. 🙁

If we add some extra `td` to the `tbody` on [this website](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/tfoot), the `thead` and `tfoot` will have empty unstyled spaces just like what we have in our hummingbird table.

WinnyChang committed:
Add A11y attributes to the table container

WinnyChang committed:
Adjust table container and focus outline; align caption to start on small screens

WinnyChang committed:
Add tabindex=0 to table containers only when scrollable

WinnyChang committed:
Force-keep 'colspan' attribute on <td> and <th> elements during sanitization

WinnyChang commented:
> @WinnyChang I think it makes sense to have the footer rows complete with no spanning applied as seen below; ![image](https://private-user-images.githubusercontent.com/5203639/461237885-9dc44095-e700-41ac-92de-1fba62e9c269.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE5Mzc4NjIsIm5iZiI6MTc1MTkzNzU2MiwicGF0aCI6Ii81MjAzNjM5LzQ2MTIzNzg4NS05ZGM0NDA5NS1lNzAwLTQxYWMtOTJkZS0xZmJhNjJlOWMyNjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDhUMDExOTIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWZlYTZkNjdhYzhmMmMyM2IyYjBiNzFiMzlmNzVmMTBjZjRlZjNiMTUxYzg1MTVlZTY1YjhjMTJkODU1ODk0NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BQNRu0NhMUwREj7E89yGDjA55SXMEN9tSijl6vrGhi4)

I've fixed the colspan issue and updated the screenshots and screen recordings!

WinnyChang committed:
Relax URI allowlist to avoid stripping valid attributes; remove forceKeepAttr hook for 'colspan'

rtibbles commented on packages/kolibri-common/components/SafeHTML/index.js:
```diff
@@ -1,7 +1,7 @@
 import './style.scss';
 
-const ALLOWED_URI_REGEXP = /^blob:https?:/i;
+const ALLOWED_URI_REGEXP = /^(?:(?:blob:https?|data):|[^a-z]|[a-z+.-]+(?:[^a-z+.\-:]|$))/i;
```
For reviewers, this is a modified copy of the default regex that DomPurify uses, but with the protocols restricted to `blob:http`, `blob:https`, and `data`.

akolson commented:
@coderabbitai full review

akolson commented on packages/kolibri-common/components/SafeHTML/index.js:
```diff
@@ -22,21 +22,70 @@ export default {
+    windowSizeClass: {
+      required: true,
+      type: String,
+    },
```
I think to make this more flexible, it makes more sense to change the name of this so it is more generic for example `appearanceOverrides` or `styleOverrides` . This can be an object where any styling can be passed including the the windowSizeClass. This is more expansive and will prevent prop bloat.

WinnyChang replied:
Thanks for highlighting the scalability aspect—I hadn’t fully considered that. I've updated the code accordingly!

WinnyChang committed:
Replace windowSizeClass prop with styleOverrides for scalable style control

WinnyChang committed:
Move table rendering logic to SafeHtmlTable component

akolson reviewed (changes requested):
Hi @WinnyChang! This LGTM! I'll however add a merge blocker so we can resolve arising conflicts, but we should be good after. Thanks and Great work

AllanOXDi reviewed (approved):
LGTM! Thanks @WinnyChang

akolson reviewed (approved):
Merge conflicts resolved. Approved!
